src/core/handle.ml

```diff
@@ -144,38 +220,117 @@ let handle_cmd : sig_state -> p_command -> sig_state * proof_data option =
-      let impl = Scope.get_implicitness a in
-      (* We scope the type of the declaration. *)
-      let a = scope_basic expo a in
-      (* We check that [a] is typable by a sort. *)
-      Typing.sort_type [] a;
-      (* We check that no metavariable remains. *)
-      if Basics.has_metas true a then
+    let ps = List.map (List.map fst) ps in
+    (List.fold_left (handle_open cmd.pos) ss ps, None)
+  | P_symbol(ms, st, t, ts_pe, e) ->
+    let x,xs,ao = st.elt in
+    (*   : U  := V begin..end *)
+    begin
+      match (ao, e, t, ts_pe) with
+      | (   None, Def, None  , Some _) ->
+        fatal x.pos ":= proofterm via proofscript but without type"
+      | (      _, Def, None  , None  ) ->
+        fatal x.pos ":= without definition term nor proofscript"
+      | _ -> ()
+    end;
+    let is_opaq {elt; _} =
+      match elt with
+      | P_opaq(Opaque) -> true
+      | _ -> false
+    in
+    let op = List.exists is_opaq ms in
+    (* Verify modifiers. *)
+    let (prop, expo, mstrat) = handle_modifiers ms in
+    (* We check that [x] is not already used. *)
+    if Sign.mem ss.signature x.elt then
+      fatal x.pos "Symbol [%s] already exists." x.elt;
+    let data =
+      (* Desugaring of arguments and scoping of [t]. *)
+      let p_t,t =
+        match t with
+        | Some t ->
+          let p_t = if xs = [] then t else Pos.none (P_Abst(xs, t)) in
+          let t = scope_basic expo p_t in
+          Some p_t, Some t
+        | None -> None, None
+      in
+      (* Desugaring of arguments and argument impliciteness. *)
+      let (ao, impl) =
+        match ao with
+        | None    -> (None, List.map (fun (_,_,impl) -> impl) xs)
+        | Some(a) ->
+          let a = if xs = [] then a else Pos.none (P_Prod(xs,a)) in
+          (Some(a), Scope.get_implicitness a)
+      in
+      let ao = Option.map (scope_basic expo) ao in
+      (* If a type [ao = Some a] is given, then we check that it is
+         typable by a sort and that [t] has type [a]. Otherwise, we
+             try to infer thetype of [t]. Unification goals are collected *)
+      (* Proof script *)
+      let (ts,pe) =
+        let p_end = Pos.make cmd.pos P_proof_end in
+        let _solve = Pos.make cmd.pos P_unif_solve in
         begin
-          fatal_msg "The type of [%s] has unsolved metavariables.\n" x.elt;
-          fatal x.pos "We have %s : %a." x.elt pp_term a
-        end;
-      (* Actually add the symbol to the signature and the state. *)
-      out 3 "(symb) %s : %a\n" x.elt pp_term a;
-      (Sig_state.add_symbol ss expo prop mstrat x a impl None,
-       None)
+          match p_t,ts_pe with
+          | Some p_t,None ->
+            [Pos.make cmd.pos (P_tac_refine(p_t))], p_end
+          | Some p_t,Some(ts,pe) ->
+            Pos.make cmd.pos (P_tac_refine(p_t))::ts, pe
+          | None,Some(ts,pe) -> ts,pe
+          | None,None -> [],p_end
+        end
+      in
+      let goals,sig_symbol,pdata_expo =
```

Comment:
A problem here is that many goals generated during scoping are missing. A simple solution would be to get them all by calling (Basics.get_metas true) on the type and the definition. A better one (that could be implemented later) would be that scoping add in some set the metas it generates on the fly.

Reply:
yes, this can be done in a separate PR later